### PR TITLE
Finish client-side support for partial responses.

### PR DESCRIPTION
**Pull request by deflaux**
Fixes https://github.com/googlegenomics/utils-java/issues/48

Also
* Handle another type of non-variant segment.
* Bump grpc dependency.

### CODE REVIEW

File: src/main/java/com/google/cloud/genomics/utils/grpc/GenomicsChannel.java [hunk 1]
```diff
@@ -58,22 +59,30 @@ private static ManagedChannel getGenomicsManagedChannel(List<ClientInterceptor>
...
-    // TODO https://github.com/googlegenomics/utils-java/issues/48
-    return getGenomicsManagedChannel(Collections.singletonList(interceptor));
+  public static ManagedChannel fromCreds(GoogleCredentials creds, String fields) throws SSLException {
+    List<ClientInterceptor> interceptors = new ArrayList();
+    interceptors.add(new ClientAuthInterceptor(creds.createScoped(Arrays.asList(GENOMICS_SCOPE)),
+        Executors.newSingleThreadExecutor()));
+    if (!Strings.isNullOrEmpty(fields)) {
+      Metadata headers = new Metadata();
+      Metadata.Key<String> partialResponseHeader =
+      Metadata.Key.of(PARTIAL_RESPONSE_HEADER, Metadata.ASCII_STRING_MARSHALLER);
```
Comment (dionloy) [inline src/main/java/com/google/cloud/genomics/utils/grpc/GenomicsChannel.java]: indentation needs fixing here and line below.

File: src/test/java/com/google/cloud/genomics/utils/ShardBoundaryTest.java [hunk 2]
```diff
@@ -48,13 +53,21 @@ public void testGetStrictVariantPredicate() {
...
-    Predicate<Variant> shardPredicate = ShardBoundary.getStrictVariantPredicate(start);
+    Predicate<Variant> shardPredicate = ShardBoundary.getStrictVariantPredicate(start, null);
     List<Variant> filteredVariants = Lists.newArrayList(Iterables.filter(Arrays.asList(variants), shardPredicate));
     assertEquals(4, filteredVariants.size());
     assertThat(filteredVariants, CoreMatchers.allOf(CoreMatchers.hasItems(atStartWithinExtent,
         atStartOverlapExtent, beyondStartWithinExtent, beyondOverlapExtent)));
   }
 
+  @Test
+  public void testGetStrictVariantPredicateInsufficientFields() {
```
Comment (dionloy) [inline src/test/java/com/google/cloud/genomics/utils/ShardBoundaryTest.java]: nice. thanks for thinking of this case.

**Comment by dionloy** [comment]
LGTM. Thanks!

File: src/main/java/com/google/cloud/genomics/utils/grpc/GenomicsChannel.java [hunk 3]
```diff
@@ -103,6 +125,25 @@ public static ManagedChannel fromDefaultCreds() throws SSLException, IOException
...
+
+  /**
+   * Create a new gRPC channel to the Google Genomics API, using OfflineAuth or the application
+   * default credentials.
+   *
+   * This library works with both the older and newer support for OAuth2 clients.
+   *
+   * https://developers.google.com/identity/protocols/application-default-credentials
+   *
+   * @param auth the OfflineAuth object
```
Comment (pgrosu) [inline src/main/java/com/google/cloud/genomics/utils/grpc/GenomicsChannel.java]: s/the/The

File: src/main/java/com/google/cloud/genomics/utils/grpc/GenomicsChannel.java [hunk 4]
```diff
@@ -103,6 +125,25 @@ public static ManagedChannel fromDefaultCreds() throws SSLException, IOException
...
+   * @param fields which fields to return in the partial response, or null for none.
```
Comment (pgrosu) [inline src/main/java/com/google/cloud/genomics/utils/grpc/GenomicsChannel.java]: s/which/Which

File: src/main/java/com/google/cloud/genomics/utils/grpc/GenomicsChannel.java [hunk 5]
```diff
@@ -103,6 +125,25 @@ public static ManagedChannel fromDefaultCreds() throws SSLException, IOException
...
+   * @return the ManagedChannel
```
Comment (pgrosu) [inline src/main/java/com/google/cloud/genomics/utils/grpc/GenomicsChannel.java]: s/the/The

File: src/main/java/com/google/cloud/genomics/utils/grpc/GenomicsChannel.java [hunk 6]
```diff
@@ -58,22 +59,30 @@ private static ManagedChannel getGenomicsManagedChannel(List<ClientInterceptor>
...
         .sslContext(GrpcSslContexts.forClient().ciphers(performantCiphers).build())
-        .intercept(interceptors).build();
+        .intercept(interceptors)
+        .build();
   }
 
   /**
    * Create a new gRPC channel to the Google Genomics API, using the provided credentials for auth.
    *
    * @param creds the credential
```
Comment (pgrosu) [inline src/main/java/com/google/cloud/genomics/utils/grpc/GenomicsChannel.java]: s/the/The

File: src/main/java/com/google/cloud/genomics/utils/grpc/GenomicsChannel.java [hunk 7]
```diff
@@ -58,22 +59,30 @@ private static ManagedChannel getGenomicsManagedChannel(List<ClientInterceptor>
...
+   * @param fields which fields to return in the partial response, or null for none.
```
Comment (pgrosu) [inline src/main/java/com/google/cloud/genomics/utils/grpc/GenomicsChannel.java]: s/which/Which

File: src/main/java/com/google/cloud/genomics/utils/grpc/GenomicsChannel.java [hunk 8]
```diff
@@ -58,22 +59,30 @@ private static ManagedChannel getGenomicsManagedChannel(List<ClientInterceptor>
...
    * @return the ManagedChannel
```
Comment (pgrosu) [inline src/main/java/com/google/cloud/genomics/utils/grpc/GenomicsChannel.java]: s/the/The

File: src/main/java/com/google/cloud/genomics/utils/grpc/GenomicsChannel.java [hunk 9]
```diff
@@ -85,7 +94,20 @@ public static ManagedChannel fromCreds(GoogleCredentials creds) throws SSLExcept
...
   public static ManagedChannel fromDefaultCreds() throws SSLException, IOException {
-    return fromCreds(CredentialFactory.getApplicationDefaultCredentials());
+    return fromDefaultCreds(null);
+  }
+
+  /**
+   * Create a new gRPC channel to the Google Genomics API, using the application default credentials
+   * for auth.
+   *
+   * @param fields which fields to return in the partial response, or null for none.
```
Comment (pgrosu) [inline src/main/java/com/google/cloud/genomics/utils/grpc/GenomicsChannel.java]: s/which/Which

File: src/main/java/com/google/cloud/genomics/utils/grpc/GenomicsChannel.java [hunk 10]
```diff
@@ -85,7 +94,20 @@ public static ManagedChannel fromCreds(GoogleCredentials creds) throws SSLExcept
...
+   * @return the ManagedChannel
```
Comment (pgrosu) [inline src/main/java/com/google/cloud/genomics/utils/grpc/GenomicsChannel.java]: s/the/The

File: src/main/java/com/google/cloud/genomics/utils/grpc/GenomicsChannel.java [hunk 11]
```diff
@@ -103,6 +125,25 @@ public static ManagedChannel fromDefaultCreds() throws SSLException, IOException
    */
   public static ManagedChannel fromOfflineAuth(OfflineAuth auth)
       throws IOException, GeneralSecurityException {
-    return fromCreds(auth.getCredentials());
+    return fromOfflineAuth(auth, null);
+  }
+
+  /**
+   * Create a new gRPC channel to the Google Genomics API, using OfflineAuth or the application
```
Comment (pgrosu) [inline src/main/java/com/google/cloud/genomics/utils/grpc/GenomicsChannel.java]: s/using/using either

File: src/main/java/com/google/cloud/genomics/utils/grpc/GenomicsChannel.java [hunk 12]
```diff
@@ -103,6 +125,25 @@ public static ManagedChannel fromDefaultCreds() throws SSLException, IOException
...
+   * default credentials.
+   *
+   * This library works with both the older and newer support for OAuth2 clients.
```
Comment (pgrosu) [inline src/main/java/com/google/cloud/genomics/utils/grpc/GenomicsChannel.java]: This sentence makes me want to read it several times.  How about the following:

*This library will work with both the newer and older versions of OAuth2 client-side support.*

File: src/main/java/com/google/cloud/genomics/utils/grpc/ReadStreamIterator.java [hunk 13]
```diff
@@ -35,8 +35,6 @@
  *
  * Includes complex retry logic to upon failure resume the stream at the last known good start
  * position without returning duplicate data.
```
Comment (pgrosu) [inline src/main/java/com/google/cloud/genomics/utils/grpc/ReadStreamIterator.java]: This sentence makes me want to read it multiple times.  How about the following:

*This function integrates complex retry logic, which upon encountering a failure will resume the stream at the last known valid start position, without having any duplicated data in returned stream.*

File: src/main/java/com/google/cloud/genomics/utils/grpc/VariantStreamIterator.java [hunk 14]
```diff
@@ -35,8 +35,6 @@
  *
  * Includes complex retry logic to upon failure resume the stream at the last known good start
  * position without returning duplicate data.
```
Comment (pgrosu) [inline src/main/java/com/google/cloud/genomics/utils/grpc/VariantStreamIterator.java]: This sentence makes me want to read it multiple times. How about the following:

This function integrates complex retry logic, which upon encountering a failure will resume the stream at the last known valid start position, without having any duplicated data in returned stream.

File: src/test/java/com/google/cloud/genomics/utils/ShardBoundaryTest.java [hunk 15]
```diff
@@ -75,21 +88,29 @@ public void testGetStrictReadPredicate() {
...
+
+  @Test
   public void testGetNonVariantOverlapsPredicate() {
     long start = 1000L;
 
     Variant overlapStart = Variant.newBuilder().setReferenceBases("T").addAlternateBases("A").setStart(900L).build();
     Variant overlapStartNonVariant = Variant.newBuilder().setReferenceBases("T").setStart(900L).setEnd(1005L).build();
 
-    Predicate<Variant> shardPredicate = ShardBoundary.getNonVariantOverlapsPredicate(start);
+    Predicate<Variant> shardPredicate = ShardBoundary.getNonVariantOverlapsPredicate(start, null);
```
Comment (pgrosu) [inline src/test/java/com/google/cloud/genomics/utils/ShardBoundaryTest.java]: This is good, but there is no test when the fields are provided as an non-null argument.  The function is an OR statement based on the following so this predicate would be true if for fields the start is within the bound and matches `VARIANT_FIELD_PATTERN`:
```Java
return Predicates.or(VariantUtils.IS_NON_VARIANT_SEGMENT, getStrictVariantPredicate(start, fields));
```

Comment (deflaux) [inline src/test/java/com/google/cloud/genomics/utils/ShardBoundaryTest.java]: @pgrosu Good point - added!

**Pull request merged**